# TASK 1 - ASSIGNMENT 2

* MRUTULA SURESH
* 28967828

### PACKAGES IMPORTED

    * Pandas
    * matplotlib
    * numpy
    * fuzz and process from fuzzywuzzy

In [1]:
!pip install fuzzywuzzy

You are using pip version 9.0.3, however version 10.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [2]:
import pandas as pd
import re
import numpy as np

In [3]:
# packages for fuzzy logic
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

C:\Users\mrutula1995\Anaconda3\lib\site-packages\fuzzywuzzy\fuzz.py:35: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [4]:
# read the dataframe
job = pd.read_csv('dataset1_with_error.csv')

**  preprocessing **
    * removing the trailing and leading zero
    * removing the '.' from teh end of the company names
    * removing the double spaces

In [5]:
#printing the number of columns and rows in the dataframe and viewing few of the rows 
print(job.shape)
job.head()

(25077, 11)


Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  not available    permanent  Gregory Martin International  Engineering Jobs   
1  not available    permanent  Gregory Martin International  Engineering Jobs   
2  not available    permanent  Gregory Martin International  Engineering Jobs   
3  not available    permanent  Gregory Martin International  Engineering Jobs   
4  not available    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000

In [6]:
#checking the data types 
job.dtypes

Id                   int64
Title               object
Location            object
ContractType        object
ContractTime        object
Company             object
Category            object
Salary per annum    object
SourceName          object
OpenDate            object
CloseDate           object
dtype: object

 We know we're working with 25077 observations of 11 variables
* **which features are Categorical variable?**
    * Categorical variables: <font color="blue">ContractType</font>, <font color="blue">ContractTime</font>
* **which features are Numerical variable?**
    * Continuous variables: <font color="blue">Salary per annum</font>

In [7]:
job.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25077 entries, 0 to 25076
Data columns (total 11 columns):
Id                  25077 non-null int64
Title               25077 non-null object
Location            25077 non-null object
ContractType        25077 non-null object
ContractTime        25077 non-null object
Company             21242 non-null object
Category            25077 non-null object
Salary per annum    25077 non-null object
SourceName          25077 non-null object
OpenDate            25077 non-null object
CloseDate           25077 non-null object
dtypes: int64(1), object(10)
memory usage: 2.1+ MB


### Some of the salary have ranges and others have a 'K' denoting 1000

In [8]:
# checking those records having a range of salary values
job[job['Salary per annum'].str.contains('-')]

Id                                              Title  \
647    46634923        RGN Nurse Hull Days or Nights **** per hour   
896    48082563  Senior Chef de Partie  One AA Rosette Hotel  T...   
980    49689021                                 Assessment Officer   
1062   51061201              Community Home Workers Melton Mowbray   
1141   52489008                                      RGN Care Home   
1292   55400384                    Live in Care Worker Jobs Exeter   
1525   56209766                                 Nurse Advisor COPD   
1797   58533485          Senior Embedded Controls Systems Engineer   
1822   58759123  Helpdesk / Desktop Support analystExcel, Word,...   
2051   60684270              Door to door / Field Sales Canvassers   
2118   61445890                              Mental Health Advisor   
2602   64145428                             DEPUTY MANAGER READING   
2724   64573377                 Deputy Care Manager / Care Manager   
2889   64809245  Contract Mechanical Draughtsperson (Huddersfield)   
2923   64813318  Manufacturing Engineer (Mechanical / Pumps / CNC)   
3057   65171402                                 Complaints Manager   
3186   65188754       Technical Sales Engineer (ElectroMechanical)   
3587   65879969                               Contract Salesperson   
3927   66197678  Front End Web Developer – (HTML, CSS, Javascript)   
3973   66367073                      Financial Controller (Assets)   
4705   66608673  INFRASTRUCTURE SUPPORT – OPERATIONS – CLOUD, S...   
5708   66935869         Account Manager, Account Executive  Energy   
5988   67017625                      Contracts Manager (Permanent)   
6225   67153390   Electrical Field Service Engineer x**** (Cranes)   
6270   67208324            Building Services Supervisor in Chelsea   
6710   67333923                               HVAC Design Engineer   
6977   67389276                     Lead Business Analyst, Banking   
7141   67638912  IT Sales Executive / New Business Development ...   
7296   67652461  MS SharePoint Senior Developer, R****k, Jo bur...   
7347   67652829               IT Equipment Installations Assistant   
...         ...                                                ...   
17906  68702082  Electronic / Electrical Test Engineer (Lithium...   
17983  68703232                             Commercial Team Leader   
18002  68703556             Service Desk / Desktop Support Analyst   
18174  68706043  Manager/ Unit Manager/ Store Manager/ Catering...   
18905  68715405  TECHNICAL LEAD CASPNET MVC SCRUM TDD HOSPITALI...   
18964  68716258       Project Manager  Southampton  ****K to ****K   
19270  68787221  IT Sales Executive  Accrington  ****k to ****k...   
19445  68804934                   Mechanical Estimator  Birmingham   
19548  68805804          Relocate to the US? QA Software Engineers   
20070  68834330                Field Service Connectivity Engineer   
20175  68837150                Restaurant Manager  ****AA Rosettes   
20854  68999922          INTERNAL SALES  **** Year Maternity Cover   
21064  69004198                           Helpdesk Apprenticeships   
21242  69005600            Outsource Manager / Employer Engagement   
21247  69005706  FRENCH AND SPANISH TEACHER  QTS  **** **** PER...   
21313  69006748                           Part Time Year 6 Teacher   
21734  69017589  BI Manager, SQL Server Financial ServicesHertf...   
22283  69036978  Application Support Analyst (legal sector)  Lo...   
22418  69039080          Oracle & SQL DBA  Investment Bank  London   
23111  69080469                                  Catering Engineer   
23152  69080768              Drilling Equipment Service Technician   
23207  69081257                    Principal/Senior ITS Consultant   
23365  69083476                                  Auto CAD Designer   
23557  69118140  Sales Manager Drain Cleaning & Inspection Equi...   
23619  69122549                                   Credit Assistant   
23815  69162548        

In [9]:
# replacing the range with mean value
sal_uni = job['Salary per annum'].unique()
for i in sal_uni:
    if '-' in i:
        sal_split = i.split('-')
        sal_split = list(map(float, sal_split))
        avg_sal = sum(sal_split)/len(sal_split)
        job.loc[job['Salary per annum'] == i,'Salary per annum'] = avg_sal

In [10]:
# replacing K with 1000
job['Salary per annum'] = (job['Salary per annum'].replace(r'[K]+$','',regex = True).astype(float)* job['Salary per annum'].str.extract(r'[\d\.]+([K]+)', expand=False).fillna(1).replace('K',10**3).astype(int))

** Since some of the salaries were in the form eg. 34K **
    * job.Salary_per_annum.replace(r'[K]+$','',regex = True).astype(float) strips the K
    * job.Salary_per_annum.str.extract(r'[\d\.]+([K]+)', expand=False).fillna(1).replace('K',10**3).astype(int) this converts the K into 1000 and multiplies to the original number

In [11]:
#converting the Salary_per_annum column to integer
job['Salary per annum'] = job['Salary per annum'].astype(int)

In [12]:
# Id column has unique values
job.Id.is_unique

True

In [13]:
# checking the categorical columns 
job.describe(include=['O'])

Title Location   ContractType ContractTime        Company  \
count              25077    25077          25077        25077          21242   
unique             25077      482              3            3           4879   
top     Maximo Developer       UK  not available    permanent  UKStaffsearch   
freq                   1     3996          19499        16194            248   

       Category     SourceName         OpenDate        CloseDate  
count     25077          25077            25077            25077  
unique        8             90             2203             2400  
top     IT Jobs  totaljobs.com  20130127T150000  20130526T120000  
freq       7085           5335               27               23

#### some of the sourcenames dont have a domain specified

In [14]:
job['sourcedomain'] = job.SourceName.str.extract(r'^[\w\-]+\.(.+)')

C:\Users\mrutula1995\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


#### This code removes the domain name and stores it in a different column 
        . example .com is removed from caterer.com and stores the domain in a separate column

In [15]:
job['sourcedomain'].unique()

array(['co.uk', 'com', 'catererandhotelkeeper.com', 'cimaglobal.com',
       'guardian.co.uk', 'planningresource.co.uk', 'newstatesman.com',
       'bighospitality.co.uk', 'opticianonline.net', 'telegraph.co.uk',
       'net', 'chinwag.com', 'gponline.com', 'scot.nhs.uk', nan,
       'electronicsweekly.com', 'perl.org', 'uk.com'], dtype=object)

In [16]:
# sourcenames that dont have an extension will have a null value
job[job['sourcedomain'].isnull()]

Id                                             Title  \
5663   66932999     Registered Midwives RM  Lincolnshire  Lincoln   
12871  68393935   Killer Javascript role Exclusive to Brightwater   
15379  68672352  Digital Account Manager  **** plus a Great Bonus   

                Location   ContractType   ContractTime            Company  \
5663             Lincoln      part_time  not available      The A24 Group   
12871            Belfast      full_time      permanent  Brightwater Group   
15379  South East London  not available      permanent        Blu Digital   

                        Category  Salary per annum         SourceName  \
5663   Healthcare & Nursing Jobs             40140      monashstudent   
12871                    IT Jobs             39996          jobcareer   
15379                 Sales Jobs             27504  admin@caterer.com   

              OpenDate        CloseDate sourcedomain  
5663   20130124T150000  20130223T150000          NaN  
12871  20120205T150000  20120306T150000          NaN  
15379  20120516T000000  20120715T000000          NaN

* ** Three of the sourcename dont have proper domain names **
    * monashstudent
    * jobcareer
    * admin@caterer.com	

** Those sourcenames without extention are replaced with sourcename of their respective company **

In [17]:
job.loc[job.SourceName == 'monashstudent', 'SourceName'] = job.SourceName[job.Company == job.Company[job.SourceName == 'monashstudent'].iloc[0]].value_counts().index[0]

In [18]:
job.loc[job.SourceName == 'jobcareer', 'SourceName'] = job.SourceName[job.Company == job.Company[job.SourceName == 'jobcareer'].iloc[0]].value_counts().index[0]

In [19]:
job.loc[job.SourceName == 'admin@caterer.com', 'SourceName'] = job.SourceName[job.Company == job.Company[job.SourceName == 'admin@caterer.com'].iloc[0]].value_counts().index[0]

In [20]:
# dropping the sourcedomain column 
job.drop('sourcedomain',axis = 1,inplace = True)

In [21]:
location  = job.Location.unique()

In [22]:
# locations are sorted and manually checked
sorted(location)

['Aberdeen',
 'Aberdeenshire',
 'Abingdon',
 'Accrington',
 'Addlestone',
 'Alcester',
 'Aldershot',
 'Alfreton',
 'Alton',
 'Altrincham',
 'Amersham',
 'Andover',
 'Antrim',
 'Arundel',
 'Ash Green',
 'Ashford',
 'Avon',
 'Axbridge',
 'Aylesbury',
 'Bagshot',
 'Banbury',
 'Bangor',
 'Barking',
 'Barnet',
 'Barnsley',
 'Barnstaple',
 'Barrow-In-Furness',
 'Basildon',
 'Basingstoke',
 'Bath',
 'Beaconsfield',
 'Bedford',
 'Bedfordshire',
 'Belfast',
 'Berkshire',
 'Bexhill-On-Sea',
 'Bexley',
 'Bicester',
 'Birkenhead',
 'Birmingham',
 'Blackburn',
 'Blackpool',
 'Blaenau Gwent',
 'Bognor Regis',
 'Boldon Colliery',
 'Bolton',
 'Bolton Le Sands',
 'Borehamwood',
 'Boston',
 'Bourne End',
 'Bournemouth',
 'Bracknell',
 'Bradford',
 'Braintree',
 'Brent Cross',
 'Brentwood',
 'Bridgend',
 'Bridgwater',
 'Brierley Hill',
 'Brighton',
 'Bristol',
 'Broadfield',
 'Bromley',
 'Bromsgrove',
 'Buckingham',
 'Buckinghamshire',
 'Burgess Hill',
 'Burnley',
 'Burton-On-Trent',
 'Bury',
 'Bury St. 

* ** Comparing the sorted location names we see the following syntactical errors **
    * Surrey/Surey
    * Leeds and Leads
    * oxford and oxfords
    * Reading and Reeding    
    * Nottingham and Nottinham

In [23]:
job['Location'].replace('Surey','Surrey',inplace=True)

In [24]:
job['Location'].replace('Leads','Leeds',inplace=True)

In [25]:
job['Location'].replace('Oxfords','Oxfordshire',inplace=True)

In [26]:
job['Location'].replace('Reeding','Reading',inplace=True)

In [27]:
job['Location'].replace('Nottinham','Nottingham',inplace=True)

In [28]:
job['Company'] = job['Company'].astype(str)

In [29]:
#removing the leading and trailing spaces from the company names
job['Company'] = job['Company'].map(lambda x: x.strip())

In [30]:
#removing the '.' at the end of each company name 
job['Company'] = job['Company'].map(lambda x: x.replace(".","") if re.search(r'[.]$',x) else x)
    
    

In [31]:
# replacing the double spaces with single space
job['Company'] = job['Company'].map(lambda x: re.sub(r"\s\s+"," ", x) if re.search(r'\s\s+',x) else x)

In [32]:
#converting the company names to lower case
job['Company'] = job['Company'].map(lambda x: x.lower())

In [33]:
# lmt is replaced with limited
job['Company'] = job['Company'].map(lambda x: re.sub(r"ltd","limited",x) if re.search(r'\b(ltd)$',x) else x)

In [34]:
job['Company'] = job['Company'].map(lambda x: re.sub(r"ltd","limited",x) if re.search(r'\b(ltd)\b',x) else x)

In [35]:
job['Company'] = job['Company'].map(lambda x: re.sub(r"ltd","limited",x) if re.search(r'(ltd)',x) else x)

In [36]:
# serv. is replaced with services
job['Company'] = job['Company'].map(lambda x: re.sub(r"\bserv\.\b","services",x) if re.search(r'serv\.',x) else x)

In [37]:
# 'nan' is replaced with empty
job['Company'] = job['Company'].map(lambda x: x.replace("nan","") if re.search(r'\bnan\b',x) else x)

In [38]:
job_company_uni = job.Company.unique()

In [39]:
job.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  not available    permanent  gregory martin international  Engineering Jobs   
1  not available    permanent  gregory martin international  Engineering Jobs   
2  not available    permanent  gregory martin international  Engineering Jobs   
3  not available    permanent  gregory martin international  Engineering Jobs   
4  not available    permanent  gregory martin international  Engineering Jobs   

   Salary per annum        SourceName         OpenDate        CloseDate  
0             24996  cv-library.co.uk  20121103T000000  20121203T000000  
1             30000  cv-library.co.uk  20130108T150000  20130408T150000  
2             30000  cv-library.co.uk  20130726T150000  20130924T150000  
3             27504  cv-library.co.uk  20121214T000000  20130314T000000  
4             24996  cv-library.co.uk  20131025T000000  20131224T000000

In [40]:
#### fuzzy logic is used to check the similarity between the company names and the following codes are written to correct the errors between the names such as spelling
#### mistakes, missing spaces etc

In [41]:
job_company_uni = list(job_company_uni)

In [42]:
# replacing company names having no spaces with its counterpart with space

for i in job_company_uni:
    if ' ' in i:
        i_nospace = ''.join(i.split())
        if i_nospace in job_company_uni:
            job.loc[job.Company == i_nospace, 'Company'] = i
            
            
            

In [43]:
# replacing a company name with its longer counterpart
for i in job_company_uni:
    if ' ' in i:
        subseti = set(i.split())
        for j in job_company_uni:
            if ' ' in j:
                subsetj = set(j.split())
                if i != j and subseti.issubset(subsetj):
                    job.loc[job.Company == i, 'Company'] = j
        
                

In [44]:
# * 'easywebrecruitment', 'easyweb recruitment'
for i in job_company_uni:
    if ' ' in i:
        k = i.replace(' ','')
        for j in job_company_uni:
            if i != j and k == j:
                job.loc[job.Company == j, 'Company'] = i

In [45]:
# replacing the company names having two spaces with company names having only one 
for i in job_company_uni:
    if ' ' in i:
        words = i.split()
        k =words[0]+''+words[1] 
        for j in range(2,len(words)): 
            k = k + ' ' + words[j]
        for m in job_company_uni:
            if k == m:
                job.loc[job.Company == i, 'Company'] = m 
                
            

In [46]:
# replacing the names without '&' with its counterpart with '&'
for i in job_company_uni:
    if '&' in i:
        k = i.replace('&',' ')
        subseti = set(k.split())
        for j in job_company_uni:
            if ' ' in j:
                subsetj = set(j.split())
                if i != j and subseti.issubset(subsetj):
                    job.loc[job.Company == j, 'Company'] = i
                    

In [47]:
# replacing names without \' with its counterpart having \'
for i in job_company_uni:
    temp = i
    if re.search(r'\'',temp):
        k = re.sub(r'\'',' ',temp)
        for j in job_company_uni:
            if i != j and k == j:
                job.loc[job.Company == j, 'Company'] = i

** the following company names are manually replaced **
* parkwell management consultants', 'parkwell management consultant'
* c.o.a.l it services limited', 'c.o.a.l it. services limited'
* ashton recruitment', 'aston recruitment'
* it skillfinder', 'i.t. skillfinder'
* 'prospero recruitment', 'prosper recruitment'
* '3 points resourcing' , '3 point resourcing'
* additional resources limited', 'additional resource'
* 'michael page finance','michael page financial services'
* 'easywebrecruitment.com','easyweb recruitment'
* 'bsc sales specialists', 'bsc sales specialists limited'
* 'bsc sales specialist limited','bsc sales specialists limited'




In [48]:
job.loc[job.Company == 'parkwell management consultant', 'Company'] = 'parkwell management consultants'

In [49]:
job.loc[job.Company == 'c.o.a.l it services limited', 'Company'] = 'c.o.a.l it. services limited'

In [50]:
job.loc[job.Company == 'ashton recruitment', 'Company'] = 'aston recruitment'

In [51]:
job.loc[job.Company == 'i.t. skillfinder', 'Company'] = 'it skillfinder'

In [52]:
job.loc[job.Company == '3 point resourcing', 'Company'] = '3 points resourcing'

In [53]:
job.loc[job.Company == 'additional resource', 'Company'] = 'additional resources limited'

In [54]:
job.loc[job.Company == 'michael page finance', 'Company'] = 'michael page financial services'

In [55]:
job.loc[job.Company == 'easywebrecruitment.com', 'Company'] = 'easyweb recruitment'

In [56]:
job.loc[job.Company == 'bsc sales specialists', 'Company'] = 'bsc sales specialists limited'

In [57]:
job.loc[job.Company == 'bsc sales specialist limited', 'Company'] = 'bsc sales specialists limited'

In [58]:
job_company_uni = job.Company.unique()
job_company_uni.shape

(4106,)

In [59]:
job_company_uni = list(job_company_uni)

### FUZZY LOGIC

    * used this code to make the changes to the names

In [60]:
#def get_matches(query,choices):
   # results = process.extractOne(query, choices, scorer =fuzz.ratio, score_cutoff=93)
   # return results

In [61]:
#matchlist = []
#for x in job_company_uni:
   # job_company_uni.remove(x)
   # if get_matches(x,job_company_uni) is not None:
     #   print([x] + list(get_matches(x,job_company_uni)))
      #  matchlist.append([x] + list(get_matches(x,job_company_uni)))
   # job_company_uni.append(x)

### some of the open dates werent in suitable formate for conversion into datatimeformat i.e months and days were replaced

In [62]:
job.loc[job.OpenDate == '20131803T120000' , 'OpenDate'] = '20130318T120000'
job.loc[job.OpenDate == '20132606T000000' , 'OpenDate'] = '20130626T000000'
job.loc[job.OpenDate == '20122003T150000' , 'OpenDate'] = '20120320T150000'
job.loc[job.OpenDate == '20121512T150000' , 'OpenDate'] = '20121215T150000'
job.loc[job.OpenDate == '20133004T150000' , 'OpenDate'] = '20130430T150000'
job.loc[job.OpenDate == '20131908T000000' , 'OpenDate'] = '20130819T000000'
job.loc[job.OpenDate == '20121406T120000' , 'OpenDate'] = '20120614T120000'
job.loc[job.OpenDate == '20131509T000000' , 'OpenDate'] = '20130915T000000'
job.loc[job.OpenDate == '20132901T150000' , 'OpenDate'] = '20130129T150000'
job.loc[job.OpenDate == '20132108T120000' , 'OpenDate'] = '20130821T120000'

In [63]:
job.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25077 entries, 0 to 25076
Data columns (total 11 columns):
Id                  25077 non-null int64
Title               25077 non-null object
Location            25077 non-null object
ContractType        25077 non-null object
ContractTime        25077 non-null object
Company             25077 non-null object
Category            25077 non-null object
Salary per annum    25077 non-null int32
SourceName          25077 non-null object
OpenDate            25077 non-null object
CloseDate           25077 non-null object
dtypes: int32(1), int64(1), object(9)
memory usage: 2.0+ MB


In [64]:
#checking the opendate and converting into the correct format
job['OpenDate'] = pd.to_datetime(job['OpenDate'],format = '%Y%m%dT%H%M%S').dt.strftime('%Y%m%dT%H%M%S')

In [65]:
# checking the closedate and converting into the correct format
job['CloseDate'] = pd.to_datetime(job['CloseDate'],format = '%Y%m%dT%H%M%S').dt.strftime('%Y%m%dT%H%M%S')

In [66]:
# checking if the open data is less than the close date
for i in range(len(job)):
    if job['OpenDate'][i] > job['CloseDate'][i]:
        temp = job['CloseDate'][i]
        job['CloseDate'][i] = job['OpenDate'][i]
        job['OpenDate'][i] = temp

        

C:\Users\mrutula1995\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
C:\Users\mrutula1995\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [67]:
job.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  not available    permanent  gregory martin international  Engineering Jobs   
1  not available    permanent  gregory martin international  Engineering Jobs   
2  not available    permanent  gregory martin international  Engineering Jobs   
3  not available    permanent  gregory martin international  Engineering Jobs   
4  not available    permanent  gregory martin international  Engineering Jobs   

   Salary per annum        SourceName         OpenDate        CloseDate  
0             24996  cv-library.co.uk  20121103T000000  20121203T000000  
1             30000  cv-library.co.uk  20130108T150000  20130408T150000  
2             30000  cv-library.co.uk  20130726T150000  20130924T150000  
3             27504  cv-library.co.uk  20121214T000000  20130314T000000  
4             24996  cv-library.co.uk  20131025T000000  20131224T000000

In [68]:
#This dataframe is saved as a csv file with utf-8 encoding 
job.to_csv('dataset1_solution.csv', index = False,encoding = 'utf-8')